pkg/customize/customize.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customize
+
+import (
+	"encoding/json"
+	"io"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+
+	edgeapi "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
+	"github.com/kcp-dev/edge-mc/pkg/jsonpath"
+)
+
+func Customize(logger klog.Logger, input *unstructured.Unstructured, customizer *edgeapi.Customizer, loc *schedulingv1alpha1.Location) *unstructured.Unstructured {
+	expandInput := input.GetAnnotations()[edgeapi.ParameterExpansionAnnotationKey] == "true"
+	expandCustomizer := customizer != nil && customizer.Annotations[edgeapi.ParameterExpansionAnnotationKey] == "true"
+	if customizer == nil && !expandInput {
+		return input
+	}
+	output := input.DeepCopy()
+	outputU := output.UnstructuredContent()
+	defs := Definitions{loc.GetLabels(), loc.GetAnnotations()}
```

Comment:
Here is the origin of nil pointer complaint when running the program with customizer only (without locations).